src/render.c

```diff
@@ -288,57 +351,105 @@ void epd_renderer_deinit() {
-        free(render_context.line_queues[i].buf);
-        free(render_context.feed_line_buffers[i]);
-        vSemaphoreDelete(render_context.feed_done_smphr[i]);
-        vTaskDelete(render_context.feed_tasks[i]);
-    }
-
     free(render_context.conversion_lut);
     free(render_context.line_threads);
     vSemaphoreDelete(render_context.frame_done);
 }
 
-EpdRect epd_difference_image_base(const uint8_t *to, const uint8_t *from,
-                                  EpdRect crop_to, int fb_width, int fb_height,
-                                  uint8_t *interlaced, bool *dirty_lines,
-                                  uint8_t *from_or, uint8_t *from_and) {
-    assert(from_or != NULL);
-    assert(from_and != NULL);
-    // OR over all pixels of the "from"-image
-    *from_or = 0x00;
-    // AND over all pixels of the "from"-image
-    *from_and = 0x0F;
+#ifdef RENDER_METHOD_LCD
+uint32_t epd_interlace_4bpp_line_VE(
+    const uint8_t* to,
+    const uint8_t* from,
+    uint8_t* interlaced,
+    uint8_t* col_dirtyness,
+    int fb_width
+);
+#endif
+
+/**
+ * Interlaces the lines at `to`, `from` into `interlaced`.
+ * returns `1` if there are differences, `0` otherwise.
+ */
+bool _epd_interlace_line(
+    const uint8_t* to,
+    const uint8_t* from,
+    uint8_t* interlaced,
+    uint8_t* col_dirtyness,
+    int fb_width
+) {
+    uint32_t dirty = 0;
+    int unaligned_start = 0;
+    int unaligned_len = fb_width;
+
+// use Vector Extensions with the ESP32-S3
+#ifdef RENDER_METHOD_LCD
+    // since display horizontal resolutions should be divisible by 16,
+    // only one end should be unaligned by 8 bytes (16 pixels)
+    uint32_t to_addr = (uint32_t)to;
+    int alignment_offset = (to_addr % 16);
+    if (alignment_offset > 0) {
+        unaligned_start = 0;
+        unaligned_len = (16 - alignment_offset) * 2;
+    } else {
+        unaligned_start = fb_width & (~0x1F);
+        unaligned_len = fb_width & 0x1F;
+    }
+
+    dirty = epd_interlace_4bpp_line_VE(
+        to + alignment_offset, from + alignment_offset, interlaced + alignment_offset * 2,
+        col_dirtyness + alignment_offset, fb_width
+    );
+#endif
+    assert((unaligned_start + unaligned_len <= fb_width));
+    for (int x = unaligned_start; x < unaligned_start + unaligned_len; x++) {
+        uint8_t t = *(to + x / 2);
+        uint8_t f = *(from + x / 2);
+        t = (x % 2) ? (t >> 4) : (t & 0x0f);
+        f = (x % 2) ? (f >> 4) : (f & 0x0f);
+        col_dirtyness[x / 2] |= (t ^ f) << (4 * (x % 2));
+        dirty |= (t ^ f);
+        interlaced[x] = (t << 4) | f;
+    }
+
+    return dirty > 0;
+}
 
-    uint8_t* dirty_cols = calloc(epd_width(), 1);
-    assert (dirty_cols != NULL);
+EpdRect epd_difference_image_base(
+    const uint8_t* to,
+    const uint8_t* from,
+    EpdRect crop_to,
+    int fb_width,
+    int fb_height,
+    uint8_t* interlaced,
+    bool* dirty_lines,
+    uint8_t* col_dirtyness
+) {
+    assert(fb_width % 16 == 0);
```

Comment:
[Two of the Color filter Array epapers](https://github.com/vroland/epdiy/blob/s3_color_implementation/src/displays.c#L94) that I'm testing in the s3_color_implementation branch do not pass this assertion ( WIDTH % 16 == 8 ) commenting that assert ends up displaying an image with pixel ghosting on the left side. This won't happen in the **main** branch

Reply:
Indeed, I need to extend both the diff and LUT processing to work on unaligned data / 8 pixel aligned data.